Auth: Add Azure/Entra workload identity support

mehighlow opened the pull request:
This PR adds support for Azure AD/Entra ID Workload Identity and federated credentials for Grafana authentication.

Fixes #78249

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

mehighlow committed:
fixes/adds azure workload identity authentication. Issue #78249

mehighlow committed:
Updates default values. Adds `workload_identity_token_file` defaults

mehighlow committed:
Updates example config. Adds `workload_identity_token_file`

mehighlow committed:
Updates docummentation: adds Federated credentials for Workload Identity

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=104807) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=104807) before we can accept your contribution.<br/><hr/>**Mykhailo Zahlada** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=104807) it.</sub>

mehighlow commented:
Hey @mgyongyosi ,
Is there a chance this could be squeezed into the next release? Our security policies are tightening, and SP secret is no longer an option for us.

mgyongyosi commented:
> Hey @mgyongyosi ,
> Is there a chance this could be squeezed into the next release? Our security policies are tightening, and SP secret is no longer an option for us.

Hey @mehighlow,

Thanks for your contribution, we are planning to review your changes soon (probably next week).

mgyongyosi reviewed (commented):
Thanks again @mehighlow! I left some suggestions that should be addressed before we can merge this!

mehighlow committed:
Update docs/sources/setup-grafana/configure-security/configure-authentication/azuread/index.md

Co-authored-by: Misi <mgyongyosi@users.noreply.github.com>

mehighlow committed:
Update docs/sources/setup-grafana/configure-security/configure-authentication/azuread/index.md

Co-authored-by: Misi <mgyongyosi@users.noreply.github.com>

mehighlow committed:
Docs: add link to official documentation. Clarifies example.

mehighlow committed:
1. Add workload_identity_enabled and workload_identity_token_file settings to [auth.azuread] for workload identity support.

2. Extend OAuthInfo struct to include workload identity fields.
3. Update OAuth authentication logic to handle Azure AD workload identity using federated token as client assertion.
4. Update sample configuration and documentation for new settings.

mehighlow committed:
ensure environment variable overrides are respected for OAuth SSO settings

- Ensure that settings loaded in pkg/services/ssosettings/strategies/oauth_strategy.go correctly reflect environment variable overrides, matching Grafana's config behavior.
- Align config loading logic with main config loader to prevent issues where INI values would override environment variables.

mehighlow committed:
updates documentation

mehighlow commented on docs/sources/setup-grafana/configure-security/configure-authentication/azuread/index.md:
```diff
@@ -65,25 +65,59 @@ To enable the Azure AD/Entra ID OAuth, register your application with Entra ID.
+            - Description: Grafana OAuth.
+            - Audience: The audience value that must appear in the external token. For Public cloud, it would be `api://AzureADTokenExchange`. See mentioned documentation for the full list of available audiences.
+
+        1. You must have set `workload_identity_enable` (env var `GF_AZURE_WORKLOAD_IDENTITY_ENABLED`) under `[auth.azuread]` to `true` in the Grafana server configuration for this to work.
```
You’re right — I mistakenly piggybacked on the [Azure Monitor Workload Identity](https://github.com/grafana/grafana/blob/main/docs/sources/setup-grafana/configure-grafana/_index.md#workload_identity_enabled) setting.

I’ve fixed that. Now there are two workload_identity_enabled settings:
	•	One for Entra ID / Azure AD integration, which should be specified under [auth.azuread]
	•	Another for Azure Monitor Workload Identity, which remains under [azure]

mehighlow committed:
test: add workload identity configuration tests for Azure AD OAuth strategy. Add test coverage for workload_identity_enabled and workload_identity_token_file settings

mehighlow commented on conf/defaults.ini:
```diff
@@ -812,6 +812,7 @@ client_id = some_client_id
 client_secret =
 managed_identity_client_id =
 federated_credential_audience =
+workload_identity_token_file = /var/run/secrets/azure/tokens/azure-identity-token
```
done

mehighlow committed:
updates documentation

mehighlow committed:
Simplify OAuth flow by removing unnecessary switch-case structure

mehighlow committed:
feat: add workload identity support to Azure AD SSO configuration UI

mehighlow commented:
@​mgyongyosi, thanks for your review and comments. I believe I’ve addressed everything and rebased my branch. Could you please take another look?

mgyongyosi committed:
Small changes

mgyongyosi commented on docs/sources/setup-grafana/configure-security/configure-authentication/azuread/index.md:
```diff
@@ -65,19 +65,57 @@ To enable the Azure AD/Entra ID OAuth, register your application with Entra ID.
+
+       1. You must have set `client_authentication` under `[auth.azuread]` to `managed_identity` in the Grafana server configuration for this to work.
+
+       1. You may optionally set `workload_identity_token_file` (env var `GF_AUTH_AZUREAD_WORKLOAD_IDENTITY_TOKEN_FILE`) under `[auth.azuread]` to `/var/run/secrets/azure/tokens/azure-identity-token` in the Grafana server configuration. This setting defaults to `/var/run/secrets/azure/tokens/azure-identity-token`.
```
I think this is not needed here, because the workload identity token file is only used when the `client_authentication` is set to `workload_identity`. Could you please remove it?

mgyongyosi commented on docs/sources/setup-grafana/configure-security/configure-authentication/azuread/index.md:
```diff
@@ -65,19 +65,57 @@ To enable the Azure AD/Entra ID OAuth, register your application with Entra ID.
+          - Description: Grafana OAuth.
+          - Audience: The audience value that must appear in the external token. For Public cloud, it would be `api://AzureADTokenExchange`. See mentioned documentation for the full list of available audiences.
+
+       1. You must have set `workload_identity_enable` (env var `GF_AUTH_AZUREAD_WORKLOAD_IDENTITY_ENABLED`) under `[auth.azuread]` to `true` in the Grafana server configuration for this to work.
```
If the `workload_identity_enabled` setting is removed then please remove this line as well.

mgyongyosi commented on docs/sources/setup-grafana/configure-security/configure-authentication/azuread/index.md:
```diff
@@ -65,19 +65,57 @@ To enable the Azure AD/Entra ID OAuth, register your application with Entra ID.
+
+       1. You must have set `client_authentication` (env var `GF_AUTH_AZUREAD_CLIENT_AUTHENTICATION`) under `[auth.azuread]` to `workload_identity` in the Grafana server configuration for this to work.
+
+       1. You may have set `workload_identity_token_file` (env var `GF_AUTH_AZUREAD_WORKLOAD_IDENTITY_TOKEN_FILE`) under `[auth.azuread]` to `/var/run/secrets/azure/tokens/azure-identity-token` in the Grafana server configuration for this to work. (Optional, defaults to `/var/run/secrets/azure/tokens/azure-identity-token`)
```
Based on the implementation, having the `workload_identity_token_file` set when `client_authenticaton = workload_identity` is a must.
```suggestion
       1. You must have set `workload_identity_token_file` (env var `GF_AUTH_AZUREAD_WORKLOAD_IDENTITY_TOKEN_FILE`) under `[auth.azuread]` to `/var/run/secrets/azure/tokens/azure-identity-token` in the Grafana server configuration for this to work. (Optional, defaults to `/var/run/secrets/azure/tokens/azure-identity-token`)
```

mehighlow replied:
Since `workload_identity_token_file` is already defined in defaults.ini, I believe that makes it optional, just in case. I don’t really expect it to change. Would you agree?

mgyongyosi reviewed (commented):
Thanks! I left some suggestions for removing the (seem to be) unnecessary `workload_identity_enabled` configuration setting.

mgyongyosi committed:
Lint + i18n gen

mehighlow committed:
refactor: remove redundant workload_identity_enabled setting as auth method gets defined by client_authentication

mehighlow committed:
refactor: remove redundant workload_identity_enabled setting as auth method gets defined by client_authentication

mehighlow committed:
update documentation

mehighlow commented on conf/defaults.ini:
```diff
@@ -812,6 +812,8 @@ client_id = some_client_id
 client_secret =
 managed_identity_client_id =
 federated_credential_audience =
+workload_identity_enabled = false
```
You’re right, `workload_identity_enabled` makes no sense, since `client_authentication` already defines the authentication method. I’ve `removed workload_identity_enabled`.

mehighlow committed:
updates documentation - configuration options table: adds `client_authentication`, `workload_identity_token_file`, and `federated_credential_audience`

mgyongyosi committed:
Small changes, lint, i18n

mgyongyosi commented:
Thanks @mehighlow for your contribution! 💯

marybaldwin commented:
@mehighlow - could you create a what's new post for this? https://admin.grafana.com/whats-new/ - we're trying to get them all by EOD tomorrow. Thank you!

marybaldwin commented:
@mgyongyosi @sarahleejane -  could you create a what's new post for this? https://admin.grafana.com/whats-new/ - we're trying to get them all by EOD tomorrow. Thank you!
